modal/_runtime/user_code_imports.py

```diff
@@ -91,11 +193,36 @@ def construct_webhook_callable(
         raise InvalidError(f"Unrecognized web endpoint type {webhook_config.type}")
 
 
+def snapshot_callback(
+    container_io_manager: "modal._runtime.container_io_manager.ContainerIOManager", function_def: api_pb2.Function
+):
+    if function_def.is_checkpointing_function and os.environ.get("MODAL_ENABLE_SNAP_RESTORE") == "1":
+        container_io_manager.memory_snapshot()
+
+
+def create_breakpoint_wrapper(container_io_manager: "modal._runtime.container_io_manager.ContainerIOManager"):
+    # Install hooks for interactive functions.
+    def breakpoint_wrapper():
+        # note: it would be nice to not have breakpoint_wrapper() included in the backtrace
+        container_io_manager.interact(from_breakpoint=True)
+        import pdb
+
+        current_frame = inspect.currentframe()
+        if current_frame is not None:
+            frame = current_frame.f_back
+            pdb.Pdb().set_trace(frame)
+        else:
+            raise RuntimeError("No current frame found")
+
+    sys.breakpointhook = breakpoint_wrapper
+
+
 @dataclass
 class ImportedFunction(Service):
     app: modal.app._App
     service_deps: Optional[Sequence["modal._object._Object"]]
     user_cls_instance = None
+    _function_def: api_pb2.Function
```

Comment:
Given that `ImportedFunction` itself is private, I'm okay with this being just:

```suggestion
    function_def: api_pb2.Function
```